File: tablet/Resources/task/QoLBasic.js
// QoLBasic.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "qolbasic",
    fieldlist = dbcommon.standardTaskFields();

fieldlist.push(
    {name: 'tto', type: DBCONSTANTS.TYPE_REAL},
    {name: 'rs', type: DBCONSTANTS.TYPE_REAL}
);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function QoLBasic(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(QoLBasic, taskcommon.BaseTask);
lang.extendPrototype(QoLBasic, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: QoLBasic,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    // OTHER

    // Standard task functions
    isComplete: function () {
        return (this.tto !== null && this.rs !== null);
    },

    getSummary: function () {
        var qol_tto = (this.tto === null) ? null : this.tto / 10,
            qol_rs = (this.rs === null) ? null : this.rs / 100;
        return (
            L("qolbasic_tto_q_s") + " " +
            lang.toFixedOrNull(this.tto, 2) + ". " +
            L("qolbasic_rs_q_s") + " " +
            lang.toFixedOrNull(this.rs, 1) + ". " +
            L("qolbasic_mean_qol") + " " +
            lang.toFixedOrNull(lang.mean(qol_tto, qol_rs), 3) +
            this.isCompleteSuffix()
        );
    },

    getDetail: function () {
        return this.getSummary();
    },

    edit: function (readOnly) {
        var self = this,
            Questionnaire = require('questionnaire/Questionnaire'),
            pages,
            questionnaire;

        pages = [
            {
                title: L("qolbasic_tto_title"),
                elements: [
                    { type: "QuestionText", text: L("qolbasic_tto_q") },
                    {
                        type: "QuestionSlider",
                        field: "tto",
                        min: 0,
                        max: 10,
                        showCurrentValueNumerically: true,
                        numberFormatDecimalPlaces: 1,
                        labels: [
                            { left: 0, text: "0" },
                            { center: "10%", text: "1" },
                            { center: "20%", text: "2" },
                            { center: "30%", text: "3" },
                            { center: "40%", text: "4" },
                            { center: "50%", text: "5" },
                            { center: "60%", text: "6" },
                            { center: "70%", text: "7" },
                            { center: "80%", text: "8" },
                            { center: "90%", text: "9" },
                            { right: 0, text: "10" }
                        ]
                    }
                ]
            },
            {
                title: L("qolbasic_rs_title"),
                elements: [
                    { type: "QuestionText", text: L("qolbasic_rs_q") },
                    {
                        type: "QuestionSlider",
                        field: "rs",
                        min: 0,
                        max: 100,
                        showCurrentValueNumerically: true,
                        numberFormatDecimalPlaces: 0,
                        labels: [
                            { left: 0, text: L("qolbasic_rs_0") },
                            { right: 0, text: L("qolbasic_rs_100") }
                        ]
                    }
                ]
            }
        ];

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn
        });
        questionnaire.open();
    }

});

module.exports = QoLBasic;

File: server/tasks/qolbasic.py
#!/usr/bin/env python3
# qolbasic.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

import pythonlib.rnc_web as ws
from cc_modules.cc_constants import (
    CTV_DICTLIST_INCOMPLETE,
)
from cc_modules.cc_html import (
    answer,
    identity,
    tr,
)
from cc_modules.cc_lang import mean
from cc_modules.cc_string import WSTRING
from cc_modules.cc_task import Task


# =============================================================================
# QoL-Basic
# =============================================================================

DP = 3


class QolBasic(Task):
    tablename = "qolbasic"
    shortname = "QoL-Basic"
    longname = "Quality of Life: basic assessment"
    fieldspecs = [
        dict(name="tto", cctype="FLOAT", min=0, max=10,
             comment="Time trade-off (QoL * 10). Prompt: ... Indicate... the "
             "number of years in full health [0-10] that you think is of "
             "equal value to 10 years in your current health state."),
        dict(name="rs", cctype="FLOAT", min=0, max=100,
             comment="Rating scale (QoL * 100). Prompt: Mark the point on the "
             "scale [0-100] that you feel best illustrates your current "
             "quality of life."),
    ]

    TASK_FIELDS = [x["name"] for x in fieldspecs]

    def get_trackers(self):
        return [
            {
                "value": self.get_tto_qol(),
                "plot_label": "Quality of life: time trade-off",
                "axis_label": "TTO QoL (0-1)",
                "axis_min": 0,
                "axis_max": 1,
            },
            {
                "value": self.get_rs_qol(),
                "plot_label": "Quality of life: rating scale",
                "axis_label": "RS QoL (0-1)",
                "axis_min": 0,
                "axis_max": 1,
            },
        ]

    def get_clinical_text(self):
        if not self.is_complete():
            return CTV_DICTLIST_INCOMPLETE
        tto_qol = self.get_tto_qol()
        rs_qol = self.get_rs_qol()
        mean_qol = mean([tto_qol, rs_qol])
        return [{
            "content":  "Quality of life: time trade-off {}, rating scale {},"
                        " mean {}.".format(
                            ws.number_to_dp(tto_qol, DP),
                            ws.number_to_dp(rs_qol, DP),
                            ws.number_to_dp(mean_qol, DP)
                        )
        }]

    def get_summaries(self):
        return [
            self.is_complete_summary_field(),
            dict(name="tto_qol", cctype="FLOAT",
                 value=self.get_tto_qol(),
                 comment="Quality of life (0-1), from time trade-off method"),
            dict(name="rs_qol", cctype="FLOAT",
                 value=self.get_rs_qol(),
                 comment="Quality of life (0-1), from rating scale method"),
        ]

    def is_complete(self):
        return (
            self.are_all_fields_complete(QolBasic.TASK_FIELDS)
            and self.field_contents_valid()
        )

    def get_tto_qol(self):
        return self.tto / 10 if self.tto is not None else None

    def get_rs_qol(self):
        return self.rs / 100 if self.rs is not None else None

    def get_task_html(self):
        tto_qol = self.get_tto_qol()
        rs_qol = self.get_rs_qol()
        mean_qol = mean([tto_qol, rs_qol])
        h = """
            <div class="summary">
                <table class="summary">
        """
        h += self.get_is_complete_tr()
        h += tr("Mean QoL", answer(ws.number_to_dp(mean_qol, DP, default=None),
                                   formatter_answer=identity))
        h += """
                </table>
            </div>
            <div class="explanation">
                Quality of life (QoL) has anchor values of 0 (none) and 1
                (perfect health), and can be asked about in several ways.
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="33%">Scale</th>
                    <th width="33%">Answer</th>
                    <td width="33%">Implied QoL</th>
                </tr>
        """
        h += tr(WSTRING("qolbasic_tto_q_s"),
                answer(ws.number_to_dp(self.tto, DP, default=None)),
                answer(ws.number_to_dp(tto_qol, DP, default=None),
                       formatter_answer=identity))
        h += tr(WSTRING("qolbasic_rs_q_s"),
                answer(ws.number_to_dp(self.rs, DP, default=None)),
                answer(ws.number_to_dp(rs_qol, DP, default=None),
                       formatter_answer=identity))
        h += """
            </table>
        """
        return h
